src/stan/mcmc/hmc/nuts/base_nuts.hpp

```diff
@@ -171,72 +141,75 @@ namespace stan {
 
             double h = this->hamiltonian_.H(this->z_);
             if (boost::math::isnan(h))
               h = std::numeric_limits<double>::infinity();
 
-            util.criterion = util.log_u + (h - util.H0) < this->max_delta_;
-            if (!util.criterion) ++(this->divergent_);
+            if ((h - H0) > this->max_deltaH_) this->divergent_ = true;
+
+            double pi = exp(H0 - h);
+            sum_weight += pi;
+            sum_metro_prob += pi > 1 ? 1 : pi;
 
-            util.sum_prob += std::min(1.0, std::exp(util.H0 - h));
-            util.n_tree += 1;
+            z_propose = this->z_;
+            rho += this->z_.p;
 
-            return (util.log_u + (h - util.H0) < 0);
+            return !this->divergent_;
 
-          } else {
+        } else {
           // General recursion
-          Eigen::VectorXd left_subtree_rho(rho.size());
-          left_subtree_rho.setZero();
           ps_point z_init(this->z_);
 
-          int n1 = build_tree(depth - 1, left_subtree_rho, &z_init,
-                              z_propose, util, writer);
+          Eigen::VectorXd rho_subtree(rho.size());
+          rho_subtree.setZero();
 
-          if (z_init_parent) *z_init_parent = z_init;
+          // Build the left subtree
+          double sum_weight_left = 0;
 
-          if (!util.criterion) return 0;
+          bool valid_left
+            = build_tree(depth - 1, rho_subtree, z_propose,
+                         H0, sign, n_leapfrog,
+                         sum_weight_left, sum_metro_prob, writer);
 
-          Eigen::VectorXd right_subtree_rho(rho.size());
-          right_subtree_rho.setZero();
-          ps_point z_propose_right(z_init);
+          sum_weight += sum_weight_left;
+          if (!valid_left) return false;
 
-          int n2 = build_tree(depth - 1, right_subtree_rho, 0,
-                              z_propose_right, util, writer);
+          // Build the right subtree
+          ps_point z_propose_right(this->z_);
+          double sum_weight_right = 0;
 
-          double accept_prob = static_cast<double>(n2) /
-            static_cast<double>(n1 + n2);
+          bool valid_right
+            = build_tree(depth - 1, rho_subtree, z_propose_right,
+                         H0, sign, n_leapfrog,
+                         sum_weight_right, sum_metro_prob, writer);
 
-          if ( util.criterion && (this->rand_uniform_() < accept_prob) )
-            z_propose = z_propose_right;
+          sum_weight += sum_weight_right;
+          if (!valid_right) return false;
 
-          Eigen::VectorXd& subtree_rho = left_subtree_rho;
-          subtree_rho += right_subtree_rho;
-
-          rho += subtree_rho;
-
-          util.criterion &= compute_criterion(z_init, this->z_, subtree_rho);
+          // Multinomial sample from right subtree
+          double accept_prob = sum_weight_right / sum_weight;
+          if (this->rand_uniform_() < accept_prob)
+            z_propose = z_propose_right;
```

Comment:
I take it `z_propose` is equal to the `z_propose_left` at this point so everything happens online?  This is very slick.